// @grafana/faro-react/dist/cjs/utils/reactVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isReactVersionAtLeast16 = exports.isReactVersionAtLeast17 = exports.isReactVersionAtLeast18 = exports.isReactVersionAtLeast19 = exports.reactVersionMajor = exports.reactVersion = void 0;
exports.getMajorReactVersion = getMajorReactVersion;
exports.isReactVersionAtLeast = isReactVersionAtLeast;
const react_1 = require("react");
exports.reactVersion = react_1.version;
exports.reactVersionMajor = getMajorReactVersion();
exports.isReactVersionAtLeast19 = isReactVersionAtLeast(19);
exports.isReactVersionAtLeast18 = isReactVersionAtLeast(18);
exports.isReactVersionAtLeast17 = isReactVersionAtLeast(17);
exports.isReactVersionAtLeast16 = isReactVersionAtLeast(16);
function getMajorReactVersion() {
    const major = exports.reactVersion.split('.');
    try {
        return major[0] ? parseInt(major[0], 10) : null;
    }
    catch (_err) {
        return null;
    }
}
function isReactVersionAtLeast(version) {
    return exports.reactVersionMajor === null ? false : exports.reactVersionMajor >= version;
}
