// @grafana/faro-react/dist/esm/profiler/FaroProfiler.js
// React is required in scope for JSX transformation with the classic transform
// @ts-expect-error - TS6133: React appears unused but is required for JSX
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, { Component } from 'react';
import { VERSION } from '@grafana/faro-web-sdk';
import { api, internalLogger } from '../dependencies';
export class FaroProfiler extends Component {
    get isOtelInitialized() {
        return !!(api === null || api === void 0 ? void 0 : api.isOTELInitialized());
    }
    get otel() {
        return api === null || api === void 0 ? void 0 : api.getOTEL();
    }
    get tracer() {
        var _a;
        return (_a = this.otel) === null || _a === void 0 ? void 0 : _a.trace.getTracer('@grafana/faro-react', VERSION);
    }
    createSpan(spanName, options) {
        var _a, _b;
        const span = this.tracer.startSpan(spanName, {
            startTime: options === null || options === void 0 ? void 0 : options.startTime,
            attributes: Object.assign({ 'react.component.name': this.props.name }, ((_a = options === null || options === void 0 ? void 0 : options.attributes) !== null && _a !== void 0 ? _a : {})),
        });
        (_b = this.otel) === null || _b === void 0 ? void 0 : _b.trace.setSpan(this.otel.context.active(), span);
        if (options === null || options === void 0 ? void 0 : options.endTime) {
            span.end(options.endTime);
        }
        return span;
    }
    createChildSpan(spanName, parent, options) {
        var _a;
        let span;
        (_a = this.otel) === null || _a === void 0 ? void 0 : _a.context.with(this.otel.trace.setSpan(this.otel.context.active(), parent), () => {
            span = this.createSpan(spanName, options);
        });
        return span;
    }
    constructor(props) {
        super(props);
        this.mountSpan = undefined;
        this.renderSpan = undefined;
        this.updateSpan = undefined;
        if (this.isOtelInitialized) {
            this.mountSpan = this.createSpan('componentMount');
        }
        else {
            internalLogger === null || internalLogger === void 0 ? void 0 : internalLogger.error('The Faro React Profiler requires tracing instrumentation. Please enable it in the "instrumentations" section of your config.');
        }
    }
    componentDidMount() {
        if (this.isOtelInitialized && this.mountSpan) {
            // Let OTel stamp the end via its hrTime helper (`performance.timeOrigin +
            // performance.now()`), which is monotonic. Passing `Date.now()` here would mix the
            // wall clock with the mount span's start time (auto-stamped by OTel from hrTime),
            // making `endTime - startTime` drift with NTP / DST / manual clock changes.
            this.mountSpan.end();
            // Start the `componentRender` span immediately. It covers the live, mounted lifetime
            // of the component and is ended in `componentWillUnmount`. By starting and ending it
            // without explicit timestamps, both boundaries are stamped from OTel hrTime
            // (monotonic). This avoids the wall-clock issues of the previous implementation
            // which captured `Date.now()` here and subtracted it from another `Date.now()` at
            // unmount — a problem that compounded over a long-lived component (router/app
            // shells can live for the entire session).
            this.renderSpan = this.createChildSpan('componentRender', this.mountSpan);
        }
    }
    shouldComponentUpdate({ updateProps }) {
        if (this.isOtelInitialized && this.mountSpan && updateProps !== this.props.updateProps) {
            const changedProps = Object.keys(updateProps).filter((key) => updateProps[key] !== this.props.updateProps[key]);
            if (changedProps.length > 0) {
                this.updateSpan = this.createChildSpan('componentUpdate', this.mountSpan, {
                    attributes: {
                        'react.component.changed_props': changedProps,
                    },
                });
            }
        }
        return true;
    }
    componentDidUpdate() {
        if (this.isOtelInitialized && this.updateSpan) {
            this.updateSpan.end();
            this.updateSpan = undefined;
        }
    }
    componentWillUnmount() {
        if (this.isOtelInitialized && this.renderSpan) {
            // End via OTel hrTime (monotonic). See `componentDidMount` for rationale.
            this.renderSpan.end();
            this.renderSpan = undefined;
        }
    }
    render() {
        return this.props.children;
    }
}
