// @grafana/faro-react/dist/cjs/errorBoundary/FaroErrorBoundary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FaroErrorBoundary = void 0;
// React is required in scope for JSX transformation with the classic transform
// @ts-expect-error - TS6133: React appears unused but is required for JSX
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const react_1 = require("react");
const faro_web_sdk_1 = require("@grafana/faro-web-sdk");
const dependencies_1 = require("../dependencies");
const utils_1 = require("../utils");
const const_1 = require("./const");
class FaroErrorBoundary extends react_1.Component {
    constructor(props) {
        super(props);
        this.state = const_1.faroErrorBoundaryInitialState;
        this.resetErrorBoundary = this.resetErrorBoundary.bind(this);
    }
    getErrorWithComponentStack(error, errorInfo) {
        if (!utils_1.isReactVersionAtLeast17 || !errorInfo.componentStack) {
            return error;
        }
        const newError = new Error(error.message);
        newError.name = `React ErrorBoundary ${error.name}`;
        newError.stack = errorInfo.componentStack;
        return newError;
    }
    static getDerivedStateFromError(error) {
        return {
            hasError: true,
            error,
        };
    }
    componentDidCatch(error, errorInfo) {
        var _a, _b, _c, _d;
        const errorWithComponentStack = this.getErrorWithComponentStack(error, errorInfo);
        (_b = (_a = this.props).beforeCapture) === null || _b === void 0 ? void 0 : _b.call(_a, errorWithComponentStack);
        dependencies_1.api.pushError(errorWithComponentStack, this.props.pushErrorOptions);
        (_d = (_c = this.props).onError) === null || _d === void 0 ? void 0 : _d.call(_c, errorWithComponentStack);
        this.setState({ hasError: true, error });
    }
    componentDidMount() {
        var _a, _b;
        (_b = (_a = this.props).onMount) === null || _b === void 0 ? void 0 : _b.call(_a);
    }
    componentWillUnmount() {
        var _a, _b;
        (_b = (_a = this.props).onUnmount) === null || _b === void 0 ? void 0 : _b.call(_a, this.state.error);
    }
    resetErrorBoundary() {
        var _a, _b;
        (_b = (_a = this.props).onReset) === null || _b === void 0 ? void 0 : _b.call(_a, this.state.error);
        this.setState(const_1.faroErrorBoundaryInitialState);
    }
    render() {
        if (!this.state.hasError) {
            return (0, faro_web_sdk_1.isFunction)(this.props.children) ? this.props.children() : this.props.children;
        }
        const element = !(0, faro_web_sdk_1.isFunction)(this.props.fallback)
            ? this.props.fallback
            : this.props.fallback(this.state.error, this.resetErrorBoundary);
        if ((0, react_1.isValidElement)(element)) {
            return element;
        }
        if (this.props.fallback) {
            dependencies_1.internalLogger === null || dependencies_1.internalLogger === void 0 ? void 0 : dependencies_1.internalLogger.warn('ErrorBoundary\n', 'Cannot get a valid ReactElement from "fallback"');
        }
        return null;
    }
}
exports.FaroErrorBoundary = FaroErrorBoundary;
