// @grafana/faro-web-tracing/dist/cjs/instrumentation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TracingInstrumentation = void 0;
const api_1 = require("@opentelemetry/api");
const core_1 = require("@opentelemetry/core");
const instrumentation_1 = require("@opentelemetry/instrumentation");
const resources_1 = require("@opentelemetry/resources");
const sdk_trace_web_1 = require("@opentelemetry/sdk-trace-web");
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const faro_web_sdk_1 = require("@grafana/faro-web-sdk");
const faroMetaAttributesSpanProcessor_1 = require("./faroMetaAttributesSpanProcessor");
const faroTraceExporter_1 = require("./faroTraceExporter");
const getDefaultOTELInstrumentations_1 = require("./getDefaultOTELInstrumentations");
const sampler_1 = require("./sampler");
const semconv_1 = require("./semconv");
// the providing of app name here is not great
// should delay initialization and provide the full Faro config,
// taking app name from it
class TracingInstrumentation extends faro_web_sdk_1.BaseInstrumentation {
    constructor(options = {}) {
        super();
        this.options = options;
        this.name = '@grafana/faro-web-tracing';
        this.version = faro_web_sdk_1.VERSION;
    }
    initialize() {
        var _a, _b, _c, _d, _e;
        const options = this.options;
        const attributes = {};
        if (this.config.app.name) {
            attributes[semantic_conventions_1.ATTR_SERVICE_NAME] = this.config.app.name;
        }
        if (this.config.app.namespace) {
            attributes[semconv_1.ATTR_SERVICE_NAMESPACE] = this.config.app.namespace;
        }
        if (this.config.app.version) {
            attributes[semantic_conventions_1.ATTR_SERVICE_VERSION] = this.config.app.version;
        }
        if (this.config.app.environment) {
            attributes[semconv_1.ATTR_DEPLOYMENT_ENVIRONMENT_NAME] = this.config.app.environment;
            /**
             * @deprecated will be removed in the future and has been replaced by ATTR_DEPLOYMENT_ENVIRONMENT_NAME (deployment.environment.name)
             * We need to keep this for compatibility with some internal services for now.
             */
            attributes[semantic_conventions_1.SEMRESATTRS_DEPLOYMENT_ENVIRONMENT] = this.config.app.environment;
        }
        const browserMeta = this.metas.value.browser;
        if ((0, faro_web_sdk_1.isArray)(browserMeta === null || browserMeta === void 0 ? void 0 : browserMeta.brands)) {
            attributes[semconv_1.ATTR_BROWSER_BRANDS] = browserMeta.brands.map((entry) => entry.brand);
        }
        if (browserMeta === null || browserMeta === void 0 ? void 0 : browserMeta.language) {
            attributes[semconv_1.ATTR_BROWSER_LANGUAGE] = browserMeta.language;
        }
        if (typeof (browserMeta === null || browserMeta === void 0 ? void 0 : browserMeta.mobile) === 'boolean') {
            attributes[semconv_1.ATTR_BROWSER_MOBILE] = Boolean(browserMeta.mobile);
        }
        if (browserMeta === null || browserMeta === void 0 ? void 0 : browserMeta.os) {
            attributes[semconv_1.ATTR_BROWSER_PLATFORM] = browserMeta.os;
        }
        if (browserMeta === null || browserMeta === void 0 ? void 0 : browserMeta.userAgent) {
            attributes[semantic_conventions_1.ATTR_USER_AGENT_ORIGINAL] = browserMeta.userAgent;
        }
        attributes[semconv_1.ATTR_PROCESS_RUNTIME_NAME] = 'browser';
        attributes[semconv_1.ATTR_PROCESS_RUNTIME_VERSION] = (_a = this.metas.value.browser) === null || _a === void 0 ? void 0 : _a.userAgent;
        attributes[semconv_1.ATTR_TELEMETRY_DISTRO_NAME] = 'faro-web-sdk';
        attributes[semconv_1.ATTR_TELEMETRY_DISTRO_VERSION] = faro_web_sdk_1.VERSION;
        Object.assign(attributes, options.resourceAttributes);
        const resource = (0, resources_1.defaultResource)().merge((0, resources_1.resourceFromAttributes)(attributes));
        const provider = new sdk_trace_web_1.WebTracerProvider({
            resource,
            sampler: {
                shouldSample: () => {
                    return {
                        decision: (0, sampler_1.getSamplingDecision)(this.api.getSession()),
                    };
                },
            },
            spanProcessors: [
                (_b = options.spanProcessor) !== null && _b !== void 0 ? _b : new faroMetaAttributesSpanProcessor_1.FaroMetaAttributesSpanProcessor(new sdk_trace_web_1.BatchSpanProcessor(new faroTraceExporter_1.FaroTraceExporter({ api: this.api }), {
                    scheduledDelayMillis: TracingInstrumentation.SCHEDULED_BATCH_DELAY_MS,
                    maxExportBatchSize: 30,
                }), this.metas),
            ],
        });
        provider.register({
            propagator: (_c = options.propagator) !== null && _c !== void 0 ? _c : new core_1.W3CTraceContextPropagator(),
            contextManager: options.contextManager,
        });
        const { propagateTraceHeaderCorsUrls, fetchInstrumentationOptions, xhrInstrumentationOptions } = (_d = this.options.instrumentationOptions) !== null && _d !== void 0 ? _d : {};
        (0, instrumentation_1.registerInstrumentations)({
            instrumentations: (_e = options.instrumentations) !== null && _e !== void 0 ? _e : (0, getDefaultOTELInstrumentations_1.getDefaultOTELInstrumentations)({
                ignoreUrls: this.getIgnoreUrls(),
                propagateTraceHeaderCorsUrls,
                fetchInstrumentationOptions,
                xhrInstrumentationOptions,
            }),
        });
        this.api.initOTEL(api_1.trace, api_1.context);
    }
    getIgnoreUrls() {
        return this.transports.transports.flatMap((transport) => transport.getIgnoreUrls());
    }
}
exports.TracingInstrumentation = TracingInstrumentation;
TracingInstrumentation.SCHEDULED_BATCH_DELAY_MS = 1000;
