// @grafana/faro-web-sdk/dist/cjs/instrumentations/_internal/activityWindowTracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityWindowTracker = void 0;
exports.isRequestStartMessage = isRequestStartMessage;
exports.isRequestEndMessage = isRequestEndMessage;
const faro_core_1 = require("@grafana/faro-core");
const const_1 = require("./monitors/const");
function isRequestStartMessage(msg) {
    return msg.type === const_1.MESSAGE_TYPE_HTTP_REQUEST_START;
}
function isRequestEndMessage(msg) {
    return msg.type === const_1.MESSAGE_TYPE_HTTP_REQUEST_END;
}
/**
 * Tracks events in a time‑boxed activity window. When the window goes quiet for `inactivityMs`,
 * it enters a draining phase: new short events are ignored; only active operations are awaited
 * until they end or `drainTimeoutMs` elapses.
 */
class ActivityWindowTracker extends faro_core_1.Observable {
    constructor(eventsObservable, options) {
        var _a, _b, _c, _d;
        super();
        this._tracking = false;
        this.eventsObservable = eventsObservable;
        this._options = {
            inactivityMs: (_a = options === null || options === void 0 ? void 0 : options.inactivityMs) !== null && _a !== void 0 ? _a : 100,
            drainTimeoutMs: (_b = options === null || options === void 0 ? void 0 : options.drainTimeoutMs) !== null && _b !== void 0 ? _b : 10 * 1000,
            isOperationStart: (_c = options === null || options === void 0 ? void 0 : options.isOperationStart) !== null && _c !== void 0 ? _c : (() => undefined),
            isOperationEnd: (_d = options === null || options === void 0 ? void 0 : options.isOperationEnd) !== null && _d !== void 0 ? _d : (() => undefined),
        };
        this._initialize();
    }
    _initialize() {
        this.eventsObservable
            .filter(() => {
            return this._tracking;
        })
            .subscribe((event) => {
            var _a, _b, _c;
            // Monotonic clock — used only for the duration delta in `stopTracking()`.
            // Not exposed as an absolute timestamp anywhere.
            this._lastEventTime = (0, faro_core_1.monoNow)();
            (_a = this._currentEvents) === null || _a === void 0 ? void 0 : _a.push(event);
            const startKey = this._options.isOperationStart(event);
            if (startKey) {
                (_b = this._activeOperations) === null || _b === void 0 ? void 0 : _b.set(startKey, true);
            }
            const endKey = this._options.isOperationEnd(event);
            if (endKey) {
                (_c = this._activeOperations) === null || _c === void 0 ? void 0 : _c.delete(endKey);
            }
            this._scheduleInactivityCheck();
        });
    }
    startTracking() {
        if (this._tracking) {
            return;
        }
        this._tracking = true;
        // Use a monotonic clock so the resulting `duration` (computed in `stopTracking()`)
        // is immune to wall-clock adjustments (NTP step, manual change, DST) that may
        // occur during the tracking window.
        this._startTime = (0, faro_core_1.monoNow)();
        this._lastEventTime = this._startTime;
        this.notify({
            message: 'tracking-started',
        });
        this._currentEvents = [];
        this._activeOperations = new Map();
        this._scheduleInactivityCheck();
    }
    stopTracking() {
        this._tracking = false;
        this._clearTimer(this._inactivityTid);
        this._clearTimer(this._drainTid);
        let duration;
        if (this.hasActiveOperations()) {
            duration = (0, faro_core_1.monoNow)() - this._startTime;
        }
        else {
            duration = this._lastEventTime ? this._lastEventTime - this._startTime : 0;
        }
        this.notify({
            message: 'tracking-ended',
            events: this._currentEvents,
            duration: duration,
        });
    }
    _scheduleInactivityCheck() {
        this._inactivityTid = startTimeout(this._inactivityTid, () => {
            if (this.hasActiveOperations()) {
                this._startDrainTimeout();
            }
            else {
                this.stopTracking();
            }
        }, this._options.inactivityMs);
    }
    _startDrainTimeout() {
        this._drainTid = startTimeout(this._drainTid, () => {
            this.stopTracking();
        }, this._options.drainTimeoutMs);
    }
    hasActiveOperations() {
        return !!this._activeOperations && this._activeOperations.size > 0;
    }
    _clearTimer(id) {
        if (id) {
            clearTimeout(id);
        }
    }
}
exports.ActivityWindowTracker = ActivityWindowTracker;
function startTimeout(timeoutId, cb, delay) {
    if (timeoutId) {
        clearTimeout(timeoutId);
    }
    //@ts-expect-error for some reason vscode is using the node types
    timeoutId = setTimeout(() => {
        cb();
    }, delay);
    return timeoutId;
}
